src/interpreter.c

```diff
@@ -20,8 +20,98 @@ typedef struct {
 
+// Backtrace support;
+#ifdef _OS_LINUX_
+extern uintptr_t __start_jl_enter_interpreter_frame;
+extern uintptr_t __stop_jl_enter_interpreter_frame;
+
+extern uintptr_t __start_jl_interpreter_frame;
+extern uintptr_t __stop_jl_interpreter_frame;
+
+#define SECT_INTERP JL_SECTION("jl_interpreter_frame")
+#define SECT_ENTER_INTERP JL_SECTION("jl_enter_interpreter_frame")
+
+#elif defined(_OS_DARWIN_)
+extern uintptr_t __start_jl_enter_interpreter_frame_val __asm("section$start$__TEXT$__jeif");
+uintptr_t __start_jl_enter_interpreter_frame = (uintptr_t)&__start_jl_enter_interpreter_frame_val;
+extern uintptr_t __stop_jl_enter_interpreter_frame_val __asm("section$end$__TEXT$__jeif");
+uintptr_t __stop_jl_enter_interpreter_frame = (uintptr_t)&__stop_jl_enter_interpreter_frame_val;
+
+extern uintptr_t __start_jl_interpreter_frame_val __asm("section$start$__TEXT$__jif");
+uintptr_t __start_jl_interpreter_frame = (uintptr_t)&__start_jl_interpreter_frame_val;
+extern uintptr_t __stop_jl_interpreter_frame_val __asm("section$end$__TEXT$__jif");
+uintptr_t __stop_jl_interpreter_frame = (uintptr_t)&__stop_jl_interpreter_frame_val;
+
+#define SECT_INTERP JL_SECTION("__TEXT,__jif")
+#define SECT_ENTER_INTERP JL_SECTION("__TEXT,__jeif")
+#else
+JL_DLLEXPORT int jl_is_interpreter_frame(uintptr_t ip)
+{
+    return 0;
+}
+
+JL_DLLEXPORT int jl_is_enter_interpreter_frame(uintptr_t ip)
+{
+    return 0;
+}
+
+#define SECT_INTERP
+#define SECT_ENTER_INTERP
+#define NO_INTERP_BT
+#endif
+
+#ifndef NO_INTERP_BT
+JL_DLLEXPORT int jl_is_interpreter_frame(uintptr_t ip)
+{
+    return __start_jl_interpreter_frame <= ip && ip <= __stop_jl_interpreter_frame;
+}
+
+JL_DLLEXPORT int jl_is_enter_interpreter_frame(uintptr_t ip)
+{
+    return __start_jl_enter_interpreter_frame <= ip && ip <= __stop_jl_enter_interpreter_frame;
+}
+
+JL_DLLEXPORT size_t jl_capture_interp_frame(uintptr_t *data, uintptr_t sp, size_t space_remaining)
+{
+    interpreter_state *s = (interpreter_state *)sp;
+    // Sentinel value to indicate an interpreter frame
+    data[0] = (uintptr_t)-1;
+    if (space_remaining <= 1)
+        return 0;
+    data[1] = (uintptr_t)s->src;
+    data[2] = (uintptr_t)s->ip;
+    return 2;
+}
+#else
+JL_DLLEXPORT int jl_is_interpreter_frame(uintptr_t ip)
+{
+    return 0;
+}
+
+JL_DLLEXPORT int jl_is_enter_interpreter_frame(uintptr_t ip)
+{
+    return 0;
+}
+
+JL_DLLEXPORT size_t jl_capture_interp_frame(uintptr_t *data, uintptr_t sp, size_t space_remaining)
+{
+    return 0;
+}
+#endif
+
+
+// This function is special. The unwinder looks for this function to find interpreter
+// stack frames. We rely on the compiler not spilling any extra variables to the stack
+// before the interpreter state. If that's not the case on some platform, that platform
+// will need this function written in assembly.
+void *NOINLINE SECT_ENTER_INTERP enter_interpreter_frame(void *(*callback)(interpreter_state *, void *), void *arg) {
+    interpreter_state state = {};
```

Comment:
wouldn't it be a lot faster to represent this explicitly, and teach jl_throw how to unwind it?
```
struct interpreter_state_chain {
    uintptr_t sp; // = __builtin_frame_address(0);
    jl_code_info_t src;
    uint32_t ip;
    struct interpreter_state_chain *next;
} global;
```

Reply:
You don't even need the `sp` field? (`&global` would be good enough.) I've thought about this before but don't know how reliable `sp` is. This PR also rely on that of course....

Reply:
@JeffBezanson and I thought about it, but it would mean more state to manage everywhere, which didn't seem worth the trade off.